### PR TITLE
Adds Pipeline Best Practices and Design Considerations

### PR DESCRIPTION
**Pull request by Ian-Skelton**
This PR adds a section regarding and pipeline design best practices and how to approach data pipeline design.

### CODE REVIEW

File: 05 ArchitecturalPatterns/dataPipelineBestPractices.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+* **Scheduled Loading Windows:** Schedule data extraction during off-peak hours or periods of low activity on the source system. This reduces the risk of contention and minimises the impact on the users of the system.
+
+## **Choosing the Right Tool for the Job**
+Selecting the appropriate tools and tech stack for your data pipeline is an important step. The choice of tools depends on various factors, including the type and volume of data, processing needs, the specific goals of the pipeline and the type of consumers at the pipelines destination. Here are some considerations for choosing the right tool for your data pipeline.
+
+* **Data Characteristics:** Identify the types of data you will be handling (structured, unstructured, or semi-structured), their formats and their volume and velocity.
+  
+* **Processing Requirements:** Determine whether your pipeline needs to handle real-time data streams or batch processing. Some tools are optimised for real-time analytics, while others excel in batch processing.
```
Comment (Jake-Pullen) [inline 05 ArchitecturalPatterns/dataPipelineBestPractices.md:24]: Is it worth linking the data streaming and batch processing to the relevant pages? Just for ease of use

File: 05 ArchitecturalPatterns/dataPipelineBestPractices.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+### How to implement an Idempotent Pipeline
+Generally speaking, to achieve an idempotent pipeline, every time data is curated and persisted to disk, the write process should honour the above principals. Following the below strategies will help with this.
+
+* **Use a mechanism to identify each unique record:**
+  * Use unique keys or natural keys to identify records. This will allow you to use upsert statements or other mechanisms to ensure that data is only inserted or updated if it doesn’t already exist.
+  * If no such key is present in the source data, use a form of hashing to to create your own. Most database management systems come with hashing functions out of the box.
+* **Replacing existing rows with new values from the source row (upserting using a MERGE statement):** 
+  * Using a MERGE statement allows inserting/updating/deleting in a single atomic operation by using the unique record key.
+  * For example, the below MERGE statement compares a source table to a target table based on a key column. If a target row is detected, it is updated. If a target row is not found, it is inserted.<br><br>
```
Comment (Jake-Pullen) [inline 05 ArchitecturalPatterns/dataPipelineBestPractices.md:84]: Not a fan of html tags in markdown, 2 spaces after your sentence and then newline will achieve a new line, 2 newlines will achieve a break

File: 05 ArchitecturalPatterns/dataPipelineBestPractices.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    ```
```
Comment (Jake-Pullen) [inline 05 ArchitecturalPatterns/dataPipelineBestPractices.md:85]: Can you add SQL after the ' please

**Review by Jake-Pullen — COMMENTED** [review]
Looks great and reads well, only nitpick comments from me

File: 05 ArchitecturalPatterns/dataPipelineBestPractices.md (hunk 3 above)
Comment (Ian-Skelton) [reply 05 ArchitecturalPatterns/dataPipelineBestPractices.md]: Good shout, done in commit [#d345c25](https://github.com/AiimiLtd/de-ai-onboarding/pull/4/commits/d345c25a2c43e189922b0b671b5d443cea29e079)

File: 05 ArchitecturalPatterns/dataPipelineBestPractices.md (hunk 2 above)
Comment (Ian-Skelton) [reply 05 ArchitecturalPatterns/dataPipelineBestPractices.md]: Good shout, done in commit [#d345c25](https://github.com/AiimiLtd/de-ai-onboarding/pull/4/commits/d345c25a2c43e189922b0b671b5d443cea29e079)

File: 05 ArchitecturalPatterns/dataPipelineBestPractices.md (hunk 1 above)
Comment (Ian-Skelton) [reply 05 ArchitecturalPatterns/dataPipelineBestPractices.md]: Good shout, done in commit [#d345c25](https://github.com/AiimiLtd/de-ai-onboarding/pull/4/commits/d345c25a2c43e189922b0b671b5d443cea29e079)

**Review by Jake-Pullen — APPROVED** [review]
lgtm

**Pull request merged**